nomad/structs/config/users.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package config
+
+import (
+	"errors"
+
+	"github.com/hashicorp/nomad/helper/pointer"
+)
+
+// UsersConfig configures things related to operating system users.
+type UsersConfig struct {
+	// MinDynamicUser is the lowest uid/gid for use in the dynamic users pool.
+	MinDynamicUser *int `hcl:"dynamic_user_min"`
+
+	// MaxDynamicUser is the highest uid/gid for use in the dynamic users pool.
+	MaxDynamicUser *int `hcl:"dynamic_user_max"`
+}
+
+// Copy returns a deep copy of the Users struct.
+func (u *UsersConfig) Copy() *UsersConfig {
+	if u == nil {
+		return nil
+	}
+	return &UsersConfig{
+		MinDynamicUser: pointer.Copy(u.MinDynamicUser),
+		MaxDynamicUser: pointer.Copy(u.MaxDynamicUser),
+	}
+}
+
+// Merge returns a new Users where non-empty/nil fields in the argument have
+// higher precedence.
+func (u *UsersConfig) Merge(o *UsersConfig) *UsersConfig {
+	switch {
+	case u == nil:
+		return o.Copy()
+	case o == nil:
+		return u.Copy()
+	default:
+		return &UsersConfig{
+			MinDynamicUser: pointer.Merge(u.MinDynamicUser, o.MinDynamicUser),
+			MaxDynamicUser: pointer.Merge(u.MaxDynamicUser, o.MaxDynamicUser),
+		}
+	}
+}
+
+// Equal returns whether u and o are the same.
+func (u *UsersConfig) Equal(o *UsersConfig) bool {
+	if u == nil || o == nil {
+		return u == o
+	}
+	switch {
+	case !pointer.Eq(u.MinDynamicUser, o.MinDynamicUser):
+		return false
+	case !pointer.Eq(u.MaxDynamicUser, o.MaxDynamicUser):
+		return false
+	default:
+		return true
+	}
+}
+
+// Validate whether UsersConfig is valid.
+//
+// Note that -1 is a valid value for min/max dynamic users, as this is used
+// to indicate the dynamic workload users feature should be disabled.
+func (u *UsersConfig) Validate() error {
+	if u == nil {
+		return errors.New("users must not be nil")
```

Comment:
Creating the errors as vars helps you make cleaner tests, where you check the error itself, instead of the error message